### PR TITLE
[ML] Add a "clear cache" control message to pytorch_inference

### PR DESCRIPTION
**Pull request by droberts195**
This is likely to be useful for QA performance tests once
it's exposed on the Java side.

### CODE REVIEW

File: bin/pytorch_inference/CCommandParser.cc [hunk 1]
```diff
@@ -281,8 +291,18 @@ CCommandParser::jsonToInferenceRequest(const rapidjson::Document& doc) {
...
+    case E_NumberOfAllocations:
+        return {controlMessageType, doc[NUM_ALLOCATIONS].GetInt(),
+                doc[REQUEST_ID].GetString()};
+    case E_ClearCache:
+        return {controlMessageType, 0, doc[REQUEST_ID].GetString()};
+    case E_Unknown:
+        break;
+    }
+
+    LOG_ABORT(<< "Programmatic error - incorrect validation of control message");
```
Comment (droberts195) [inline bin/pytorch_inference/CCommandParser.cc:305]: We should never get here as `validateControlMessageJson()` should have responded with an error if the intended methods are called in the correct order.

File: include/core/CCompressedLfuCache.h [hunk 2]
```diff
@@ -161,7 +161,7 @@ class CCompressedLfuCache {
                         readValue(value);
                         return;
                     }
-                    m_LostCount += lastEvictedCount;
+                    m_RemovedCount += lastEvictedCount;
```
Comment (droberts195) [inline include/core/CCompressedLfuCache.h:164]: I shifted this loss of count from `lost` to `removed` as it's similar to what happens in a resize.

File: include/core/CCompressedLfuCache.h [hunk 3]
```diff
@@ -174,6 +174,18 @@ class CCompressedLfuCache {
         return false;
     }
 
+    void clear() {
+        this->guardWrite(DONT_TIME_OUT, [&] {
+            m_NumberLookups = 0;
+            m_NumberHits = 0;
+            m_LostCount = 0;
+            m_RemovedCount = 0;
```
Comment (droberts195) [inline include/core/CCompressedLfuCache.h:182]: I decided to reset all the counters to zero here.  The alternative would be to increment the removed count instead to indicate everything being removed.  It's fairly academic anyway, as we don't expose these counters to the Java side - they are only for our unit tests at the moment.

File: bin/pytorch_inference/Main.cc [hunk 4]
```diff
@@ -170,6 +171,18 @@ void writeThreadSettings(ml::core::CJsonOutputStreamWrapper& wrappedOutputStream
...
+void writeSimpleAck(ml::core::CJsonOutputStreamWrapper& wrappedOutputStream,
+                    const std::string& requestId) {
+    ml::core::CRapidJsonConcurrentLineWriter jsonWriter(wrappedOutputStream);
+    jsonWriter.StartObject();
+    jsonWriter.Key(ACK);
+    jsonWriter.StartObject();
+    jsonWriter.Key(ml::torch::CCommandParser::REQUEST_ID);
+    jsonWriter.String(requestId);
+    jsonWriter.EndObject();
+    jsonWriter.EndObject();
```
Comment (droberts195) [inline bin/pytorch_inference/Main.cc:183]: The structure of this object highlights the benefit in hoisting `request_id` up to the top level.

I'll probably do that before the Java side of this change, so this object structure will never exist in the Java code, but didn't want to mix two changes in this PR.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Pull request merged**